Consolidating file watchers

fabiocav opened the pull request:
Relies on #1489 

Fixes issues with multiple watchers and resource utilization, which is particularly impactful when running against Azure Files.

fabiocav commented:
To better review the changes (since this builds on top of #1489) , look at commit https://github.com/Azure/azure-webjobs-sdk-script/pull/1490/commits/926ecba5fde1ad6497fe6e3c8fa31647d6264c52

brettsam commented on src/WebJobs.Script/Description/FunctionInvokerBase.cs:
```diff
@@ -97,9 +99,10 @@ protected bool InitializeFileWatcherIfEnabled()
-                _fileWatcher = new AutoRecoveringFileSystemWatcher(functionScriptDirectory);
-                _fileWatcher.Changed += OnScriptFileChanged;
+                _fileChangeSubscription = Host.EventManager.OfType<FileEvent>()
+                    .Where(f => string.Equals(f.Source, "ScriptFiles", StringComparison.Ordinal) &&
```
Move these sources to constants...

fabiocav replied:
Meant to do that before sending this! 👍

brettsam commented on src/WebJobs.Script/Host/ScriptHostManager.cs:
```diff
@@ -44,16 +47,18 @@ public class ScriptHostManager : IScriptHostEnvironment, IDisposable
+                _fileEventSource = new FileWatcherEventSource(EventManager, "ScriptFiles", config.RootScriptPath);
+
+                EventManager.OfType<FileEvent>()
+                    .Where(f => string.Equals(f.Source, "ScriptFiles", StringComparison.Ordinal))
```
Would it be useful to add a Linq extension like `WithSource(string)` since this looks like it'll be the common pattern? This is assuming `Source` moves to `ScriptEvent`, which may or may not make sense.

fabiocav replied:
We can. The source here is pretty specific to file watchers (it's part of the `FileEvent`), so hopefully, its use will be limited. We could create the extension if we feel it is needed, though.

brettsam commented on src/WebJobs.Script/Eventing/File/FileEvent.cs:
```diff
@@ -0,0 +1,21 @@
+            FileChangeArguments = args;
+        }
+
+        public string Source { get; set; }
```
Should `Source` be on `ScriptEvent`?

fabiocav replied:
This was meant to identify the source of the `FileEvent`, but I can see your point (and better understand your previous comment). The `ScriptEvent` should really only contain members that are common and required across events, but I agree that this would be a good candidate.

brettsam commented on src/WebJobs.Script/GlobalSuppressions.cs:
```diff
@@ -174,6 +174,10 @@
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.SpacingRules", "SA1009:Closing parenthesis must be spaced correctly", Justification = "<Pending>", Scope = "member", Target = "~T:Microsoft.Azure.WebJobs.Script.Description.NodeFunctionInvoker")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Scope = "member", Target = "Microsoft.Azure.WebJobs.Script.ScriptHost.#Create(Microsoft.Azure.WebJobs.Script.IScriptHostEnvironment,Microsoft.Azure.WebJobs.Script.Eventing.IScriptEventManager,Microsoft.Azure.WebJobs.Script.ScriptHostConfiguration,Microsoft.Azure.WebJob ...
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Scope = "member", Target = "Microsoft.Azure.WebJobs.Script.ScriptHostManager.#.ctor(Microsoft.Azure.WebJobs.Script.ScriptHostConfiguration,Microsoft.Azure.WebJobs.Script.Config.ScriptSettingsManager,Microsoft.Azure.WebJobs.Script.IScriptHostFactory,Microsoft.Azure.WebJ ...
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2213:DisposableFieldsShouldBeDisposed", MessageId = "_fileEventSource", Scope = "member", Target = "Microsoft.Azure.WebJobs.Script.ScriptHostManager.#Dispose(System.Boolean)")]
```
Is this required b/c StyleCop doesn't like `_fileEventSource?.Dispose()`? I ran into that too -- it's confusing.

fabiocav replied:
Yeah, StyleCop doesn't seem to recognize that pattern (or at least it seems a bit flaky)

brettsam commented on src/WebJobs.Script/Host/ScriptHostManager.cs:
```diff
@@ -43,27 +47,37 @@ public class ScriptHostManager : IScriptHostEnvironment, IDisposable
             {
-                _scriptFileWatcher = new AutoRecoveringFileSystemWatcher(config.RootScriptPath);
-                _scriptFileWatcher.Changed += OnScriptFileChanged;
+                _fileEventSource = new FileWatcherEventSource(EventManager, "ScriptFiles", config.RootScriptPath);
```
It looks like ScriptHost sets up the same `FileWatcherEventSource`. Does that mean they're both registering two watchers of the same files? Would that cause dupe events to fire?

brettsam commented on src/WebJobs.Script/Description/FunctionInvokerBase.cs:
```diff
@@ -97,9 +99,10 @@ protected bool InitializeFileWatcherIfEnabled()
-                _fileWatcher.Changed += OnScriptFileChanged;
+                _fileChangeSubscription = Host.EventManager.OfType<FileEvent>()
+                    .Where(f => string.Equals(f.Source, "ScriptFiles", StringComparison.Ordinal) &&
+                    f.FileChangeArguments.Name.StartsWith(Metadata.Name, StringComparison.OrdinalIgnoreCase))
```
_I somehow lost this comment in my review_

The previous FileWatcher matched on the script filename. This new subscription is looking at `StartsWith(functionName)`. Could that have some false matches if I had two functions named `MyFunction` and `MyFunction1`?

fabiocav replied:
Good catch. I'll make sure the function name is taken from the path (the actual directory name)

mathewc commented on src/WebJobs.Script/Description/FunctionInvokerBase.cs:
```diff
@@ -97,9 +99,11 @@ protected bool InitializeFileWatcherIfEnabled()
-                _fileWatcher = new AutoRecoveringFileSystemWatcher(functionScriptDirectory);
-                _fileWatcher.Changed += OnScriptFileChanged;
+                string functionBasePath = Metadata.Name + Path.DirectorySeparatorChar;
+                _fileChangeSubscription = Host.EventManager.OfType<FileEvent>()
```
Sweet 👍

mathewc commented on src/WebJobs.Script/Eventing/File/FileEventSources.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.Azure.WebJobs.Script.Eventing
+{
+    public static class FileEventSources
```
Wondering if you don't instead just want a general ScriptEventSources class for all the various sources, not file specific?

fabiocav commented:
@mathewc @brettsam wrapped up with the remaining changes and have addressed the feedback you guys had. Let me know if you have any additional concerns.

brettsam commented:
@fabiocav -- Not a concern, but I'm curious. It looks like we register the same `FileWatcherEventSource` twice: `new FileWatcherEventSource(EventManager, "ScriptFiles", config.RootScriptPath);` -- once in `ScriptHost` and one in `ScriptHostManager`. Do we get dupe events from those?

fabiocav commented:
@brettsam sorry, I thought I had replied to that comment earlier.

That code path in the `ScriptHostManager` is not executed when running the web host (so it doesn't run in production). I kept that in place to minimize the change but if the flag is set, this will receive the notifications from the publisher setup in the `ScriptHost`, so I'll go ahead and remove that.

Eventually, I wan to move that logic from the `ScriptHost` to the `ScriptHostManager`, executing it as part of the regular initialization so that we have a single publisher ever setup and don't need to tear it down at the host restarts.

fabiocav commented:
@brettsam replying to your earlier comment:

> If I'm reading this right -- you're removing the need for having a FileWatcher-per-Invoker. Instead, there's one that watches all files and invokers subscribe to that Observable. Is that right?
Yes, that is correct.

> Assuming so -- is there a measurable perf (or other) benefit to this change? I like it either way; just curious :-)

There will be a performance impact as there will be less change notification handles (particularly important when running against Azure Files), but the key thing here is reliability and better resource utilization. Azure Files has a limit on the number of file change notification subscriptions and in some cases (depending on the number of functions and how scaled out the app was) we are hitting that limit, which causes notification failures as well as significant performance degradation.

mathewc commented on src/WebJobs.Script/Host/ScriptHostManager.cs:
```diff
@@ -44,16 +47,18 @@ public class ScriptHostManager : IScriptHostEnvironment, IDisposable
-                _scriptFileWatcher.Changed += OnScriptFileChanged;
+                // We only setup a subscription here as the actual ScriptHost will create the publisher
+                // when initialized.
+                _fileEventSubscription = EventManager.OfType<FileEvent>()
```
Subscribing to file events will be very common in our codebase - you're adding a few in this PR. We might consider an extension method(s) to facilitate common patterns, allowing these to be simplified to EventManager.OnFileEvent(OnScriptFileChanged).

Just mentioning this as a future suggestion.

fabiocav replied:
Yes, I do like the idea.

mathewc commented on test/WebJobs.Script.Tests.Integration/Host/ScriptHostManagerTests.cs:
```diff
@@ -141,6 +149,9 @@ public ScriptHostManagerTests()
                         // rename directory & delete old blob
                         Directory.Move(oldDirectory, newDirectory);
+
+                        resetEvent.Wait(TimeSpan.FromSeconds(60));
```
60 seconds seems a bit long. We're waiting only 10 seconds above.

fabiocav replied:
Ah, yeah. I'll bring that down. I bumped it up to buy me some time while debugging some of this stuff and forgot to set it back.
